Better value decoding

jeremyrsmith opened the pull request:
This commit changes how values are decoded internally. It introduces the `ValueDecoder` trait, which is responsible for decoding a value from either text or binary using a particular strategy.

Additionally, OIDs are now used to map to the name of the postgres receive function for a type, which is a reliable way of deciding how the value should be decoded. These receive functions are then mapped to ValueDecoders.  This change allows for cusom value decoders to be specified, and also allows for the remote database to be queried for its types; custom types may be present with a non-fixed OID, but which use existing receive functions.  Thus, type handling is more robust against postgres use cases.

Changes to the user-facing API:
- Client.apply now accepts additional optional arguments:
  - `customReceiveFunctions` which allows for specifying a `PartialFunction` that maps receive function names to `ValueDecoder`s
  - `binaryResults` which allows specifying that the client should ask the server for results in binary format (which is more robust and performant than text)
- Timestamps now use appropriate `java.time` representations by default:
  - `java.time.LocalDateTime` for timestamp
  - `java.time.ZonedDateTime` for timestamp with time zone
  - `java.time.LocalDate` for date

Compatibility changes:
- Java 8 is now required (since `java.time` is used)

The integration tests have also been changed to use an environment-configurable Postgres instance, rather than a hard-coded one.  This will allow the integration tests to actually run in travis-ci.

Tests also now use ScalaCheck where possible.

There are also a few fixes for bugs which didn't become apparent until binary mode was possible.

Commit:
Better value decoding

This commit changes how values are decoded internally. It introduces the
`ValueDecoder` trait, which is responsible for decoding a value from either
text or binary using a particular strategy.

Additionally, OIDs are now used to map to the name of the postgres receive
function for a type, which is a reliable way of deciding how the value should
be decoded. These receive functions are then mapped to ValueDecoders.  This
change allows for cusom value decoders to be specified, and also allows for
the remote database to be queried for its types; custom types may be present
with a non-fixed OID, but which use existing receive functions.  Thus, type
handling is more robust against postgres use cases.

Changes to the user-facing API:
- Client.apply now accepts additional optional arguments:
  - `customReceiveFunctions` which allows for specifying a `PartialFunction` that
    maps receive function names to `ValueDecoder`s
  - `binaryResults` which allows specifying that the client should ask the server
    for results in binary format (which is more robust and performant than text)
- Timestamps now use appropriate `java.time` representations by default:
  - `java.time.LocalDateTime` for timestamp
  - `java.time.ZonedDateTime` for timestamp with time zone
  - `java.time.LocalDate` for date

Compatibility changes:
- Java 8 is now required (since `java.time` is used)

The integration tests have also been changed to use an environment-configurable
Postgres instance, rather than a hard-coded one.  This will allow the integration
tests to actually run in travis-ci.

Tests also now use ScalaCheck where possible.

There are also a few fixes for bugs which didn't become apparent until binary
mode was possible.

Commit:
Travis env configuration was wrong (uses one list item per run, not per variable)

Commit:
Add missing tags to BackendMessageParser

Fix String generator for tests

jeremyrsmith commented:
Note that the comment from @coveralls is now wrong.  Coverage is up 52.1% to 68.848% as reflected in the checks.

vkostyukov commented on .travis.yml:
```diff
@@ -6,10 +6,17 @@ scala:
   - oraclejdk8
 
+services:
+  - postgresql
```
I didn't know you can do that with TravisCI. Neat!

Commit:
Small project fixes

Rename root project to `finagle-postgres` instead of `root`
Add caching to travis

jeremyrsmith commented:
@vkostyukov Any objection to merging this PR?  The biggest deal breaker I can see is that it removes compatibility with older Java versions, but I'm not sure how much of an issue that is (do people still use JDK 7 and under?)

jeremyrsmith commented:
In light of today's announcement that Finagle is moving to JDK8-only, this PR becomes even more compelling!  Anyone?  Anyone? @vkostyukov? @travisbrown? @bueller?

Commit:
Add missing test for date type
